text/0000-crate-quarantine.md

```diff
@@ -0,0 +1,156 @@
+
+Once a crate version has been quarantined, then a member of the crates.io team will review the quarantined crate, ideally within 1 week[^sla]. Once reviewed, the crate may be either approved and published to the index, or rejected and deleted. Either way, the crate owner(s) will be notified by e-mail, assuming crates.io has a verified e-mail available for their account.
+
+Should the crate owner(s) disagree with the quarantine action, they may appeal to the moderation team.
+
+## When will a quarantine rule be added?
+
+Ultimately, this will be decided by the crates.io team on a case by case basis. There are two primary scenarios that are currently envisioned where this functionality would be used:
+
+### Large scale spam attack
+
+In the event of a large scale spam attack on the entire site, we may want to either restrict individual users (if the spam is coming from a finite set of users) or quarantine all published versions temporarily (using a `.*` crate name rule) until it subsides, at which point we can analyse which crate versions are actually valid.
+
+### Malicious users
+
+As an alternative to locking or banning a user, when we have high confidence (due to user reports or automated scanning) that a user is uploading malicious crates then a rule may be added for that specific user until the crates.io team is able to communicate with them.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The initial implementation of this RFC will only allow for rules on the crate name or a specific user. Further rule types may be added in the future by the crates.io team as required.
+
+## crates.io
+
+Most changes will need to occur on crates.io. They are described below.
+
+### Database
+
+New tables will need to be added to capture quarantine rules on a per-crate and per-user basis. Crates will be matched by matching their name against a regular expression in each rule; users will be matched by their external user ID[^uid].
+
+The `versions` table will need a new `quarantined` boolean flag, much like the existing `yanked` flag.[^state]
+
+### Publishing
+
+The crates.io database will be expanded to store and capture rules that, if matched, will cause a crate version that is being published to enter the `quarantined` state, rather than being immediately published. These rules will be defined and maintained by crates.io admins.
+
+The publish endpoint will notify the publisher (generally `cargo publish`) of the quarantined version by extending the response[^publish-response] in two ways:
+
+1. A new field will be added to the response called `quarantined`.
+2. A human readable warning will be added to the `warnings` field within the response.
+
+When a crate version is quarantined, a support ticket will also be opened within the crates.io help desk, thereby notifying the crates.io team of the newly quarantined version.
```

Comment:
If there are a large number of crates, the number of tickets could also represent a spam or denial of service attack on the help desk system. So I'd suggest:

```suggestion
When a crate version is quarantined, a support ticket will also be opened within the crates.io help desk, thereby notifying the crates.io team of the newly quarantined version. if additional crates are quarantined for that user or regex, a comment will be added to the existing open ticket. (This prevents the help desk being spammed.)
```

Reply:
I like that, thanks!

Reply:
The other alternative is updating the existing ticket description. But I don't think the difference matters much at an RFC level.

Reply:
Agreed.